File: src/java/io/FileInputStream.cpp
#include "java/io/FileInputStream.h"

namespace java {
namespace io {

FileInputStream::FileInputStream():
    stream(nullptr),
    isPipe(false),
    standardInput(false)
{
}

FileInputStream::FileInputStream(const File &file):
    stream(file.open("rb")),
    isPipe(false),
    standardInput(false)
{
}

FileInputStream::FileInputStream(const char *fileName):
    stream(nullptr),
    isPipe(false),
    standardInput(false)
{
    stream = File::openHandle(fileName, "rb");
}

FileInputStream::FileInputStream(FILE *fileHandle, bool pipeInput, bool isStandardInput):
    stream(fileHandle),
    isPipe(pipeInput ? 1 : 0),
    standardInput(isStandardInput)
{
}

FileInputStream::~FileInputStream() {
    dispose();
}

bool
FileInputStream::isOpen() const {
    return stream != nullptr;
}

bool
FileInputStream::isPipeInput() const {
    return isPipe;
}

bool
FileInputStream::isStandardInput() const {
    return standardInput;
}

int
FileInputStream::read() {
    if ( stream == nullptr ) {
        return -1;
    }
    return fgetc(stream);
}

int
FileInputStream::read(unsigned char *buffer, int offset, int length) {
    if ( stream == nullptr ) {
        return -1;
    }
    if ( buffer == nullptr || offset < 0 || length < 0 ) {
        return -1;
    }
    if ( length == 0 ) {
        return 0;
    }
    const int readCount = static_cast<int>(fread(&buffer[offset], 1, static_cast<std::size_t>(length), stream));
    if ( readCount == 0 && feof(stream) ) {
        return -1;
    }
    return readCount;
}

long
FileInputStream::tell() const {
    if ( stream == nullptr ) {
        return -1;
    }
    return ftell(stream);
}

void
FileInputStream::close() {
    if ( stream == nullptr ) {
        return;
    }
    if ( !standardInput ) {
        if ( isPipe ) {
            pclose(stream);
        } else {
            File::closeHandle(stream);
        }
    }
    stream = nullptr;
    isPipe = false;
    standardInput = false;
}

void
FileInputStream::dispose() {
    close();
}

}
}

File: typescript/src/main/java/io/FileInputStream.ts
import { InputStream } from "./InputStream";

const fs = require("node:fs");

export class FileInputStream extends InputStream {
  private fileDescriptor: number | null;
  private position: number;

  public constructor(fileName: string) {
    super();
    this.fileDescriptor = null;
    this.position = 0;
    if (fileName && fileName.length > 0) {
      try {
        this.fileDescriptor = fs.openSync(fileName, "r");
      }
      catch (_error) {
        this.fileDescriptor = null;
      }
    }
  }

  public read(): number;
  public read(buffer: Uint8Array, offset: number, length: number): number;
  public read(buffer?: Uint8Array, offset?: number, length?: number): number {
    if (this.fileDescriptor === null) {
      return -1;
    }

    if (buffer === undefined) {
      const localBuffer = new Uint8Array(1);
      const readCount = fs.readSync(this.fileDescriptor, localBuffer, 0, 1, this.position);
      if (readCount <= 0) {
        return -1;
      }
      this.position += readCount;
      return localBuffer[0];
    }

    const safeOffset = offset ?? 0;
    const safeLength = length ?? 0;
    if (safeOffset < 0 || safeLength < 0 || safeOffset + safeLength > buffer.length) {
      return -1;
    }
    if (safeLength === 0) {
      return 0;
    }

    const readCount = fs.readSync(this.fileDescriptor, buffer, safeOffset, safeLength, this.position);
    if (readCount <= 0) {
      return -1;
    }
    this.position += readCount;
    return readCount;
  }

  public close(): void {
    if (this.fileDescriptor === null) {
      return;
    }
    fs.closeSync(this.fileDescriptor);
    this.fileDescriptor = null;
  }

  public override dispose(): void {
    this.close();
  }
}
